frontend/src/lib/components/widgets/DataFrame/hooks/useCustomRenderer.ts

```diff
@@ -0,0 +1,148 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from "react"
+
+import {
+  DataEditorProps,
+  DrawCustomCellCallback,
+  Rectangle,
+  Theme as GlideTheme,
+  drawTextCell,
+} from "@glideapps/glide-data-grid"
+import { useExtraCells } from "@glideapps/glide-data-grid-cells"
+
+import {
+  BaseColumn,
+  CustomCells,
+  isMissingValueCell,
+} from "src/lib/components/widgets/DataFrame/columns"
+
+// Token used for missing values (null, NaN, etc.)
+const NULL_VALUE_TOKEN = "None"
+
+/**
+ * Draw a red indicator in the top right corner of the cell
+ * to indicate that the cell is required.
+ */
+export function drawRequiredIndicator(
+  ctx: CanvasRenderingContext2D,
+  rect: Rectangle,
+  theme: GlideTheme
+): void {
+  ctx.save()
+  ctx.beginPath()
+  ctx.moveTo(rect.x + rect.width - 8, rect.y + 1)
```

Comment:
I feel like some comments here could be useful. Why 8? Is it moving left? right?

Reply:
Added some commentary 👍